### DISCUSSION TITLE
Dynamic default value for action inputs

### DISCUSSION THREAD

**Discussion by aslisabanci** [community/community#25218] [Actions]
<p>Hi, is it possible to provide dynamic default values to action inputs, when defining an action file in action.yaml?</p>
<pre><code class="lang-auto">name: '...'
description: '...'
inputs:
  test_input:
    description: '...'
    required: true
    default: 'parametric default value here?'
</code></pre>
<p>ie. similar to how we provide<code>env</code> or <code>secrets</code> variables as <code>${{ secrets.A_SECRET }}</code> to these inputs in our workflow files, would that be possible to specify a parametric input on the action definition file instead?</p>
<p>Thank you!</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/aslisabanci">@aslisabanci</a>，</p>
<p>Dynamic default value is NOT supported for the event input. It should be hardcoded with <code>string</code> format.</p>
<p>As doc mentioned:</p>
<h4><a href="https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputsinput_iddefault" rel="noopener nofollow ugc"> <strong><code>inputs.&lt;input_id&gt;.default</code></strong> </a></h4>
<p><strong>Optional</strong>  A  <code>string</code>  representing the default value. The default value is used when an input parameter isn’t specified in a workflow file.</p>
<p>Thanks</p>

**Reply by Lovebugpaul50** [reply]
- - [ ] # > []()~~![MicrosoftProductandServicesDPA(WW)(English)(Jan022024)(CR).docx](https://github.com/community/community/files/15430403/MicrosoftProductandServicesDPA.WW.English.Jan022024.CR.docx)

~~

**Comment by aslisabanci** [comment]
<p>Okay then, I already was using the string defaults but wanted to see if there’s any other way around this.<br>
Thanks for your answer!</p>

**Comment by mrmike** [comment]
<p>As a workaround I’ve started using parameter expansion to provide default value no matter I have access to event input. I wrote quick post about it <a href="https://dev.to/mrmike/github-action-handling-input-default-value-5f2g" class="inline-onebox" rel="noopener nofollow ugc">Github Action - handling input default value - DEV Community</a></p>

**Comment by SkPhilipp** [comment]
<p>Contrary to the above answer, you can reference values from for the context in your input `default`. For example in GitHub’s own actions/checkout they’re using <code>${{ github.token }}</code> as a default for the token input in <a href="https://github.com/actions/checkout/blob/8f9e05e482293f862823fcca12d9eddfb3723131/action.yml#L24" class="inline-onebox" rel="noopener nofollow ugc">checkout/action.yml at 8f9e05e482293f862823fcca12d9eddfb3723131 · actions/checkout · GitHub</a><br>
GitHub resolves these expressions to their value when running your action.</p>

**Reply by PaulRBerg** [reply]
I tried this with `$${{ secrets.FOO }}` and it didn't work

**Reply by SkPhilipp** [reply]
Not all variables which may depend on an environment are available at every level of your workflows. Within the scope of a job you'll have secrets available but that may not be the case for configuration at a level above jobs. (Or maybe it's the double dollar notation.) So without linking the full workflow it's hard to say what a solution might be @PaulRBerg

**Reply by slominskir** [reply]
```
default: ${{ github.event.repository.name }}
```
Doesn't work either (in `workflow_dispatch` at least).  Fails with:

```
The workflow is not valid. .github/workflows/myworkflow.yml (Line: 27, Col: 18): Unrecognized named-value: 'github'. Located at position 1 within expression: github.event.repository.name
```

I resorted to no default value and then adding an additional step with bash parameter expansion and step outputs as suggested above.  However, set-output is deprecated so it looks different now.  This is a ton of boiler plate code to do what should be the one-liner default value:

```
    steps:
      -
        name: Sort out the input param and set default if needed
        id: getlabel
        run: |
          LABEL_INPUT="${{ github.event.inputs.label }}"
          echo "label=${LABEL_INPUT:-myorg/${{ github.event.repository.name }}}" >> $GITHUB_OUTPUT
      -
        echo ${{ steps.getlabel.outputs.label }}   
```

**Reply by SkPhilipp** [reply]
Do keep in mind the original question is on topic of action.yaml, not workflow definitions.

**Reply by slominskir** [reply]
We've been discussing GitHub Action defaults in general (both workflows and actions) as that's what mrmike's blog linked above is about.    The solution I posted is an updated version to mrmike's avoiding since deprecated set-output syntax and works equally well in a workflow and action context.   As far as I know, this workaround is really the only solution other than don't use dynamic values as defaults in GitHub Actions.
